File: Code Out/2024/Solutions/Steam Gift Cards/steamGiftCards.py
#!/bin/python3

import collections

#
# Complete the 'countValid' function below.
#
# The function accepts following parameters:
#  1. INTEGER n
#  2. STRING_ARRAY codes
#

def countValid(n, codes):
    count = 0
    for s in codes:
        val, code = s.split(".")
        if not val:
            count += 1
            continue
        elif len(val) == len(code):
            count += val == code
            continue
        queue = collections.deque(list(val))
        for char in code:
            if char == queue[0]:
                queue.popleft()
            if not len(queue):
                count += 1
                break
    print(count)

if __name__ == '__main__':
    n = int(input().strip())

    codes = []

    for _ in range(n):
        codes_item = input()
        codes.append(codes_item)

    countValid(n, codes)

File: Code Out/2024/Solutions/Steam Gift Cards/steamGiftCards.js
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

/*
 * Complete the 'countValid' function below.
 *
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. STRING_ARRAY codes
 */

function countValid(n, codes) {
    let count = 0;
    for (const s of codes) {
        let split = s.split('.');
        if (!split[0]) {
            count++;
            continue;
        }
        if (split[0].length == split[1].length) {
            count += split[0] == split[1];
            continue;
        }
        let queue = {};
        for (const v in split[0])
            queue[v] = split[0][v];
        let front = 0; let back = split[0].length;
        for (const char of split[1]) {
            if (char == queue[front])
                delete queue[front++];
            if (!(back - front)) {
                count++;
                break;
            }
        }
    }
    console.log(count);
}

function main() {
    const n = parseInt(readLine().trim(), 10);

    let codes = [];

    for (let i = 0; i < n; i++) {
        const codesItem = readLine();
        codes.push(codesItem);
    }

    countValid(n, codes);
}